obspy/io/sac/arrayio.py

```diff
@@ -0,0 +1,637 @@
+from .util import is_valid_enum_int, is_same_byteorder, byteswap
+
+
+def init_header_arrays(*args):
+    """
+    Initialize arbitrary header arrays.
+
+    Parameters
+    ----------
+    which : str, {'float', 'int', 'str'}
+        Specify which arrays to initialize, and desired order.
+        If omitted, returned arrays are ('float', 'int', 'str').
+
+    Returns
+    -------
+    arrays : list of numpy.ndarrays
+        The desired SAC header arrays.
+
+    """
+    if len(args) == 0:
+        args = ('float', 'int', 'str')
+    #http://stackoverflow.com/a/13052254/745557
+    out = []
+    for itype in args:
+        if itype == 'float':
+            # null float header array
+            # XXX: why not init native byte order?
+            hf = np.ndarray(70, dtype=native_str('<f4'))
+            hf.fill(HD.FNULL)
+            out.append(hf)
+        elif itype == 'int':
+            # null integer header array
+            hi = np.ndarray(40, dtype=native_str('<i4'))
+            hi.fill(HD.INULL)
+            # set logicals to 0, not -1234whatever
+            for hdr in HD.INTHDRS:
+                if hdr.startswith('l'):
+                    hi[HD.INTHDRS.index(hdr)] = 0
+            # TODO: initialize enumerated values to something?
+            # calculate distances be default
+            hi[HD.INTHDRS.index('lcalda')] = 1
+            out.append(hi)
+        elif itype == 'str':
+            # null string header array
+            hs = np.ndarray(24, dtype=native_str('|S8'))
+            hs.fill(HD.SNULL)
+            out.append(hs)
+        else:
+            raise ValueError("Unrecognized header array type {}".format(itype))
+
+    return out
+
+
+def read_sac(source, headonly=False, byteorder=None, checksize=False):
+    """
+    Read a SAC binary file.
+
+    Parameters
+    ----------
+    source : str or file-like object
+        Full path string for File-like object from a SAC binary file on disk.
+        If the latter, open 'rb'.
+    headonly : bool
+        If headonly is True, only read the header arrays not the data array.
+    byteorder : str {'little', 'big'}, optional
+        If omitted or None, automatic byte-order checking is done, starting with
+        native order. If byteorder is specified and incorrect, a SacIOError is
+        raised.
+    checksize : bool, default False
+        If true, check that theoretical file size from header matches disk size.
+
+    Returns
+    -------
+    hf, hi, hs : numpy.ndarray
+        The float, integer, and string header arrays.
+    data : numpy.ndarray or None
+        float32 data array. If headonly is True, data will be None.
+
+    Raises
+    ------
+    ValueError
+        Unrecognized byte order.
+    SacIOError
+        File not found, incorrect specified byteorder, theoretical file size
+        doesn't match header, or header arrays are incorrect length.
+
+    """
+    # TODO: rewrite using "with" statement instead of open/close management.
```

Comment:
FYI: Just using `with` is not enough if we want it to work with filenames, open files and `io.BytesIO` objects. In the long run we should implement #952. For now, I think what has been done here is a fairly simple solution to the problem: https://github.com/obspy/obspy/blob/master/obspy/io/cmtsolution/core.py